crates/llms/src/model2vec.rs

```diff
@@ -0,0 +1,202 @@
+        let Self {
+            name,
+            normalize,
+            parallelism,
+            embed_max_token_length,
+            embed_custom_batch_size,
+            ..
+        } = self;
+        write!(
+            f,
+            "Model2Vec: {name}, normalize: {normalize:?}, parallelism: {parallelism:?}, embed_max_token_length: {embed_max_token_length:?}, embed_custom_batch_size: {embed_custom_batch_size:?}"
+        )
+    }
+}
+
+#[async_trait]
+impl Embed for Model2Vec {
+    async fn embed(
+        &self,
+        input: EmbeddingInput,
+    ) -> Result<Vec<Vec<f32>>, super::embeddings::Error> {
+        self.embed_sync(input)
+    }
+
+    fn embed_sync(&self, input: EmbeddingInput) -> Result<Vec<Vec<f32>>, super::embeddings::Error> {
+        let embedding_input = match input {
+            EmbeddingInput::String(s) => vec![s],
+            EmbeddingInput::StringArray(sentences) => sentences,
+            _ => {
+                return Err(UnsupportedEmbeddingInput {
+                    model: self.name.clone(),
+                    message: "Model2Vec models only support strings or vectors of strings"
+                        .to_string(),
+                });
+            }
+        };
+
+        Ok(self.model.encode_with_args(
+            &embedding_input,
+            self.embed_max_token_length,
+            self.embed_custom_batch_size.unwrap_or(1024),
+        ))
+    }
+
+    fn supports_sync_embeddings(&self) -> bool {
+        true
+    }
+
+    fn parallelism(&self) -> Option<usize> {
+        self.parallelism
+    }
+
+    fn size(&self) -> i32 {
+        -1
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::embeddings::Embed;
+    use crate::model2vec::Model2Vec;
+    use async_openai::types::EmbeddingInput;
+
+    #[tokio::test]
+    async fn test_embed() {
+        // This embedding is dim 256
+        let model = Model2Vec::from_params(
+            "minishlab/potion-base-8M",
+            None,
+            None,
+            None,
+            None,
+            None,
+            None,
+        )
+        .expect("Must instantiate");
+
+        let embed_sentence = model
+            .embed(EmbeddingInput::String("hello world".to_string()))
+            .await;
+
+        assert!(embed_sentence.is_ok());
+
+        let embed_sentence = embed_sentence.expect("Must embed sentence");
+        assert_eq!(embed_sentence.len(), 1);
+        assert_eq!(embed_sentence[0].len(), 256);
+
+        let embed_sentences = model
+            .embed(EmbeddingInput::StringArray(vec![
+                "i can eat glass".to_string(),
+                "for it does not hurt me".to_string(),
+            ]))
+            .await;
+
+        assert!(embed_sentences.is_ok());
+
+        let embed_sentences = embed_sentences.expect("Must embed sentences");
+        assert_eq!(embed_sentences.len(), 2);
+        for embedded_sentence in embed_sentences {
+            assert_eq!(embedded_sentence.len(), 256);
```

Comment:
Given these are static models, it should be possible to snapshot their output right?